The aim is to train a neural network that will detect digits from 0-9. For training the neural network, MNIST (Modified National Institute of Standards and Technology database) dataset is used. The MNIST database of handwritten digits, has a training set of 60,000 examples, and a test set of 10,000 examples.

MNIST dataset is provided with the keras so, you can directly use it without downloading it separately.

In [2]:
# import the dataset
from keras.datasets import mnist

#load the data using load_data()
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

Using TensorFlow backend.


11493376/11490434 [==============================] - 29s 3us/step


The data is encoded as Numpy arrays. Let's look at the shape and length of these arrays.

In [17]:
print('The shape of the training data: ', train_images.shape)
print('The shape of the training labels: ', train_labels.shape)
print('The shape of the testing data: ', test_images.shape)
print('The shape of the testing labels: ', test_labels.shape)
print('The length of the training data: ', len(train_images))
print('The length of the training labels: ', len(train_labels))
print('The length of the testing data: ', len(test_images))
print('The length of the testing labels: ', len(test_labels))

The shape of the training data:  (60000, 28, 28)
The shape of the training labels:  (60000,)
The shape of the testing data:  (10000, 28, 28)
The shape of the testing labels:  (10000,)
The length of the training data:  60000
The length of the training labels:  60000
The length of the testing data:  10000
The length of the testing labels:  10000


We can see that there are 6000 images of size 28 x 28 for training and 10000 images of size 28 x 28 for testing. The shape of the train and test labels is (6000,) which means that it is a list of labels which are mapped to each image in the train and test set respectively. 

Now, let's look at the contents of label arrays.

In [18]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [19]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

It can be observed that these labels are numbers from 0-9. Now let's first build our neural network (or the model) for this classification task. Then the above data will be fed to the model and lastly, the model will be tested on the test data to see if the model is correctly trained.

In [20]:
# import models and layers to define the model
from keras import models
from keras import layers

# Sequential neural network is used
network = models.Sequential()

# let's add 2 layers to the model
network.add(layers.Dense(512, activation='relu', input_shape=(28 *28,)))
network.add(layers.Dense(10, activation='softmax'))

``` network = models.Sequential() ``` means that the model used is Sequential. Other types of models are also there. We will look into them in other tutorials.

``` network.add(layers.Dense(512, activation='relu', input_shape=(28 *28,))) ``` means that the first layer that we have added is of type Dense and it contains 512 units. They are Densely connected (also known as fully connected layers). The activation function used is relu (Rectilinear unit). The shape of the input to this layer is of the for 28 x 28

``` network.add(layers.Dense(10, activation='softmax')) ``` means that this layer contains is again of type Dense and contains 10 units. The activation function used is softmax. It is a 10 way softmax layer which means that it will return an array of 10 probabilities which will sum to 1 (each probability corresponds to each class from 0 to 9)

Now, let's compile the model by using appropriate optimizer, loss function and evaluation metric.

In [22]:
network.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])

The model is ready now. We'll just need to feed the data into the model. To feed in the data, the data needs to be reshaped in the form which is expected by the model. We'd also need to convert all the values to the closed inveral [0,1]. To convert each image in the interval 0 to 1, we'd need some information about these images. Since, these images are greyscale, the maximum value must be 255 and minimum must be 0. We can also check this from the dataset.

In [23]:
train_images = train_images.reshape((60000,28*28)) # since the input to the model is of the shape 28x28
train_images = train_images.astype('float32')/255 # model expects floating point values in 99% of the cases
test_images = test_images.reshape((10000,28*28)) # since the input to the model is of the shape 28x28
test_images = test_images.astype('float32')/255 # model expects floating point values in 99% of the cases

Let's also convert labels to binary categorical form. Keras provides an inbuilt function for doing it.

In [24]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

Now, the data se entirely ready to be fed into the neural network. We will use fit() function to train the model. The first parameter to the following function are the training images, next argument is the label list. The named argument epoch specifies the number of iterations. Batch size defines the number of samples you want to use at one go.

Note that, 1 iteration corresponds to training of all the samples not just 128 (in this case).  Batch represents the number of samples you want to use at a time. (More about it in further tutorials)

In [25]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
60000/60000 [==============================] - 21s 342us/step - loss: 0.2488 - acc: 0.9287
Epoch 2/5
60000/60000 [==============================] - 17s 288us/step - loss: 0.1027 - acc: 0.9693
Epoch 3/5
60000/60000 [==============================] - 20s 328us/step - loss: 0.0666 - acc: 0.9800
Epoch 4/5
60000/60000 [==============================] - 19s 318us/step - loss: 0.0485 - acc: 0.9857
Epoch 5/5
60000/60000 [==============================] - 20s 328us/step - loss: 0.0370 - acc: 0.9889


Two quantities can be seen after each iteration. Loss and the accuracy of the model. After 5 iterations, the model has achieved 98.89% accuracy. Now, let's see how is it performing on the test set. We use evaluate() function for evaluating it on the test set. The function returns the loss and the accuracy of the model on the test set. 

In [26]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
test_acc

10000/10000 [==============================] - 2s 224us/step


0.9799

We have achieved 97.99% accuracy on the test set. This is lower than the training accuracy. This might be due to overfitting of the model on the training set.